File: server/modules/profile/services/reserved-username.ts
/**
 * Nomes reservados / equipa e limpeza de caracteres invisíveis em username de perfil.
 *
 * Migrado de legacy/backend/models/profileUsernameReserved.ts. Regex de zero-width
 * construído por code point (mesmo motivo de `shared/utils/safe-text.ts`) em vez de
 * escape literal — evita depender de caractere invisível sobrevivendo no `.ts`.
 */
const ZERO_WIDTH_SPACE_START = 0x200b;
const ZERO_WIDTH_JOINER_END = 0x200d;
const BOM = 0xfeff;

const INVISIBLE_CHARS_PATTERN = new RegExp(
  '[' + String.fromCharCode(ZERO_WIDTH_SPACE_START) + '-' + String.fromCharCode(ZERO_WIDTH_JOINER_END) + String.fromCharCode(BOM) + ']',
  'g'
);

const RESERVED = new Set(
  [
    'admin', 'administrator', 'support', 'suporte', 'root', 'genesis', 'genesisminer',
    'genesis-miner', 'minestation', 'staff', 'moderator', 'mod', 'official', 'system',
    'equipe', 'team', 'helpdesk'
  ].map((s) => s.toLowerCase())
);

export function isReservedProfileUsername(username: string): boolean {
  const t = String(username || '').trim().toLowerCase().replace(/\s+/g, ' ');
  if (!t) return true;
  const compact = t.replace(/[\s_-]/g, '');
  if (RESERVED.has(t) || RESERVED.has(compact)) return true;
  for (const r of RESERVED) {
    if (t.startsWith(`${r} `) || t.startsWith(`${r}_`) || t.startsWith(`${r}-`)) return true;
  }
  return false;
}

export function stripInvisibleUsernameChars(raw: string): string {
  return String(raw || '').replace(INVISIBLE_CHARS_PATTERN, '');
}

File: rust/genesis-core/src/auth/reserved_username.rs
//! Reserved profile usernames — mirror Node `reserved-username.ts`.

/// Zero-width space (U+200B).
const ZERO_WIDTH_SPACE: char = '\u{200B}';
/// Zero-width joiner (U+200D) — end of ZW range with ZWSP.
const ZERO_WIDTH_JOINER: char = '\u{200D}';
/// BOM (U+FEFF).
const BOM: char = '\u{FEFF}';

/// Exact reserved tokens (lowercase) — Node `RESERVED`.
const RESERVED: &[&str] = &[
    "admin",
    "administrator",
    "support",
    "suporte",
    "root",
    "genesis",
    "genesisminer",
    "genesis-miner",
    "minestation",
    "staff",
    "moderator",
    "mod",
    "official",
    "system",
    "equipe",
    "team",
    "helpdesk",
];

fn is_invisible_username_char(c: char) -> bool {
    (c >= ZERO_WIDTH_SPACE && c <= ZERO_WIDTH_JOINER) || c == BOM
}

/// Strip zero-width / BOM chars from a profile username candidate.
pub fn strip_invisible_username_chars(raw: &str) -> String {
    raw.chars()
        .filter(|c| !is_invisible_username_char(*c))
        .collect()
}

fn reserved_has(token: &str) -> bool {
    RESERVED.iter().any(|r| *r == token)
}

/// `true` when the username is reserved / empty / prefix of a reserved token.
pub fn is_reserved_profile_username(username: &str) -> bool {
    let t = username
        .trim()
        .to_lowercase()
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    if t.is_empty() {
        return true;
    }
    let compact: String = t
        .chars()
        .filter(|c| !matches!(c, ' ' | '_' | '-'))
        .collect();
    if reserved_has(&t) || reserved_has(&compact) {
        return true;
    }
    for r in RESERVED {
        if t.starts_with(&format!("{r} "))
            || t.starts_with(&format!("{r}_"))
            || t.starts_with(&format!("{r}-"))
        {
            return true;
        }
    }
    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_exact_reserved() {
        assert!(is_reserved_profile_username("Admin"));
        assert!(is_reserved_profile_username("SUPORTE"));
    }

    #[test]
    fn rejects_compact_variants() {
        assert!(is_reserved_profile_username("genesis-miner"));
        assert!(is_reserved_profile_username("genesisminer"));
    }

    #[test]
    fn rejects_prefix_separator() {
        assert!(is_reserved_profile_username("admin_oficial"));
        assert!(is_reserved_profile_username("root-user"));
    }

    #[test]
    fn rejects_empty() {
        assert!(is_reserved_profile_username(""));
    }

    #[test]
    fn accepts_normal() {
        assert!(!is_reserved_profile_username("jogador123"));
    }

    #[test]
    fn strips_invisible() {
        let zwsp = ZERO_WIDTH_SPACE;
        let bom = BOM;
        assert_eq!(
            strip_invisible_username_chars(&format!("a{zwsp}b{bom}c")),
            "abc"
        );
        assert_eq!(strip_invisible_username_chars("jogador 123"), "jogador 123");
    }
}
